flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java

```diff
@@ -0,0 +1,170 @@
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        for (RecordReader<KeyValue> reader : sortedReaders) {
+            reader.close();
+        }
+    }
+
+    private void initElements() throws IOException {
+        minHeap =
+                new PriorityQueue<>(
+                        (e1, e2) -> {
+                            int result =
+                                    userKeyComparator.compare(
+                                            e1.iterator.current().key(),
+                                            e2.iterator.current().key());
+                            if (result != 0) {
+                                return result;
+                            }
+                            return Long.compare(
+                                    e1.iterator.current().sequenceNumber(),
+                                    e2.iterator.current().sequenceNumber());
+                        });
+        for (int i = 0; i < sortedReaders.size(); i++) {
+            RecordIterator<KeyValue> iterator = sortedReaders.get(i).readBatch();
+            if (iterator != null && iterator.advanceNext()) {
+                minHeap.offer(new Element(iterator, i));
+            }
+        }
+    }
+
+    private void supplyElements() throws IOException {
+        RecordReader<KeyValue> reader = sortedReaders.get(topElement.index);
+        RecordIterator<KeyValue> iterator = reader.readBatch();
+        if (iterator != null && iterator.advanceNext()) {
+            minHeap.offer(new Element(iterator, topElement.index));
+        }
+    }
+
+    /** The iterator iterates on {@link SortMergeReader}. */
+    private class SortMergeIterator implements RecordIterator<KeyValue> {
+
+        private boolean innerIteratorsAdvanceAhead = true;
+
+        @Override
+        public boolean advanceNext() throws IOException {
+            if (minHeap.size() > 0) {
+                if (innerIteratorsAdvanceAhead) {
+                    // keep pace with the inner iterators
+                    innerIteratorsAdvanceAhead = false;
+                    topElement = minHeap.peek();
+                    return true;
+                }
+                topElement = minHeap.poll();
+                // find the top iterator to advance
+                RecordIterator<KeyValue> topIterator = topElement.iterator;
+                if (topIterator.advanceNext()) {
+                    minHeap.offer(new Element(topIterator, topElement.index));
+                    topElement = minHeap.peek();
+                    return true;
+                } else {
+                    // reach the end of current batch, should invoke the next batch
+                    return false;
+                }
+            } else if (topElement != null) {
+                // keep pace with the rest long tail iterator
+                return topElement.iterator.advanceNext();
```

Comment:
Why here? should not end of reader?

Reply:
If we offer all elements to the heap, there's no need to do this. (For better code readability) However, if there's one long-tail reader left and we don't choose to offer the element to the heap, this check is needed I think.